content/python/hardware_components.py.md

```diff
@@ -0,0 +1,172 @@
+            if 'hardwareGenericComponentModel' in component['hardwareComponentModel']:
+                capacity = component['hardwareComponentModel']['hardwareGenericComponentModel']['capacity']
+                component_type = component['hardwareComponentModel']['hardwareGenericComponentModel'][
+                    'hardwareComponentType']['type']
+                table.add_row([component['hardwareComponentModel']['id'],
+                               capacity,
+                               component['hardwareComponentModel']['manufacturer'],
+                               component['hardwareComponentModel']['name'],
+                               component['hardwareComponentModel']['version'],
+                               component_type
+                               ])
+
+        print(table)
+
+    def list_hardware_component_firmware(self, server_id, hardware_component_id):
+        mask = 'filteredMask[activeComponents[hardwareComponentModel[firmwares]]]'
+        object_filter = {
+            "hardware": {
+                "id": {
+                    "operation": server_id
+                },
+                "activeComponents": {
+                    "hardwareComponentModel": {
+                        "id": {
+                            "operation": hardware_component_id
+                        }
+                    }
+                }
+            }
+        }
+
+        result = self.client['SoftLayer_Account'].getHardware(mask=mask, filter=object_filter)
+
+        table = PrettyTable()
+        table.title = "Hardware Component Firmware"
+        table.field_names = ['id', 'createDate', 'hardwareComponentModelId', 'version']
+
+        list_firmware = list()
+        for active_component in result:
+            for component in active_component['activeComponents']:
+                if 'firmwares' in component['hardwareComponentModel']:
+                    for firmware in component['hardwareComponentModel']['firmwares']:
+                        list_firmware.append(firmware)
+
+        list_result = sorted(list_firmware, key=lambda x: x['createDate'])
+        last_firmware = list_result[len(list_result) - 1]
+        table.add_row([last_firmware['id'],
+                       last_firmware['createDate'],
+                       last_firmware['hardwareComponentModelId'],
+                       last_firmware['version'],
+                       ])
+
+        print(table)
+
+    def update_firmware(self, hardware_id, ipmi=0, raidController=0, bios=0, harddrive=0):
+        result = self.client['SoftLayer_Hardware_Server'].createFirmwareUpdateTransaction(ipmi, raidController, bios,
+                                                                                          harddrive, id=hardware_id)
+
+        if result:
+            print("Update firmware successfully")
+        else:
+            print("Update firmware failed")
+
+
+if __name__ == "__main__":
+    main = HardwareComponents()
+
+    hardware_id = 1234
```

Comment:
keep in mind add an example of how to get to `hardware_id`, (See Contributing [Requirements 2.)](https://github.com/softlayer/githubio_source/blob/master/CONTRIBUTING.md#requirements).

Reply:
Instead of using a specific Id here (which a customer might not know offhand), use a hostname or IP address and use that to look up the id, as I think that makes this example a lot more useful to people who just want to copy/paste code and have it work.

Reply:
Done.